### PR TITLE
interpolation added for TVP.

### PR DESCRIPTION
**Pull request by bhuvanmdev**
# What does this PR do?

Adds interpolation to the tvp model for the position and pad embeddings.

Addresses #30579 

## Who can review?
@amyeroberts

### CODE REVIEW

File: src/transformers/models/tvp/modeling_tvp.py [hunk 1]
```diff
@@ -183,34 +183,94 @@ def __init__(self, config):
...
+            align_corners=False,
+        )
+        embedding = embedding.permute(0, 2, 3, 1)  # (batch_size, height, width, hidden_dim)
+        return embedding
+
+    def add_2d_positional_embeddings(self, grid, interpolate_pos_encoding: bool = False):
         """
         Args:
             grid: (batch_size, height, width, hidden_dim)
+            interpolate_pos_encoding: (bool) Whether to interpolate the pre-trained position encodings.
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:219]:
```suggestion
            interpolate_pos_encoding: (`bool`, *optional*, defaults to `False`):
                Whether to interpolate the pre-trained position encodings.
```

File: src/transformers/models/tvp/modeling_tvp.py [hunk 2]
```diff
@@ -183,34 +183,94 @@ def __init__(self, config):
...
-        row_shape = (1,) * (len(grid.shape) - 3) + (height, 1, hidden_dim)  # (1, height, 1, hidden_dim)
-        grid = grid + row_position_embeddings.view(*row_shape)  # broadcast automatically
-
+        # (height, )
+        row_position_ids = torch.arange(
+            min(self.max_grid_row_position_embeddings, height), dtype=torch.long, device=grid.device
+        )
+        # (height, hidden_dim)
+        row_position_embeddings = self.row_position_embeddings(row_position_ids)
+        # (1, height, 1, hidden_dim)
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:232]: I realise this size comment comes from the original code, but (1, height, 1, hidden_dim)` doesn't match what is happening here, unless `(len(grid.shape) - 3)` is equal to 1, in which case this is pointless

File: src/transformers/models/tvp/modeling_tvp.py [hunk 3]
```diff
@@ -183,34 +183,94 @@ def __init__(self, config):
...
         """
         batch_size, height, width, hidden_dim = grid.shape
 
         # add row-wise position embeddings
-        row_position_ids = torch.arange(height, dtype=torch.long, device=grid.device)  # (height, )
-        row_position_embeddings = self.row_position_embeddings(row_position_ids)  # (height, hidden_dim)
-        row_shape = (1,) * (len(grid.shape) - 3) + (height, 1, hidden_dim)  # (1, height, 1, hidden_dim)
-        grid = grid + row_position_embeddings.view(*row_shape)  # broadcast automatically
-
+        # (height, )
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:226]: I owuld tidy this up to make it clearer 

```py
        row_height = min(self.max_grid_row_position_embeddings, height)
        row_position_ids = torch.arange(row_height, dtype=torch.long, device=grid.device)

        # (height, hidden_dim)
        row_position_embeddings = self.row_position_embeddings(row_position_ids)
        # (1, height, 1, hidden_dim)
        row_shape = (1,) * (len(grid.shape) - 3) + (row_height, 1, hidden_dim)
        # (1, height, 1, hidden_dim)
        row_position_embeddings = row_position_embeddings.view(*row_shape)
        
        # add column-wise position embeddings
        row_width = min(self.max_grid_col_position_embeddings, width)
        col_position_ids = torch.arange(row_width, dtype=torch.long, device=grid.device)
        # (width, hidden_dim)
        col_position_embeddings = self.col_position_embeddings(col_position_ids)
        col_shape = (batch_size, 1, row_width, hidden_dim)
        # (batch_size, 1, width, hidden_dim)
        col_position_embeddings = col_position_embeddings.view(*col_shape)
        # (batch_size, height, width, hidden_dim)
        positional_embeddings = row_position_embeddings + col_position_embeddings
        
        # This interpolation gets triggered ONLY when the input image dim is larger in any dimenstion than the original position embeddings
        if interpolate_pos_encoding and (
            height > self.max_grid_row_position_embeddings or width > self.max_grid_row_position_embeddings
        ):
            grid = grid + self.interpolate_pos_encoding(positional_embeddings, height, width)
        else:
            grid = grid + positional_embeddings
        return grid
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on adding this!

File: src/transformers/models/tvp/modeling_tvp.py (hunk 2 above)
Comment (bhuvanmdev) [reply src/transformers/models/tvp/modeling_tvp.py:232]: Then what should I change the comment to?
should I..
1. remove it completely.
2. keep it as it is, but mention `assuming grid.shape == 4 i.e (batch_size, height, width, hidden_dim)`.

File: tests/models/tvp/test_modeling_tvp.py [hunk 4]
```diff
@@ -265,3 +265,37 @@ def test_inference_with_head(self):
...
+        input_ids = torch.tensor([[1, 2]])
+        attention_mask = torch.tensor([[1, 1]])
+        encoding.update({"input_ids": input_ids, "attention_mask": attention_mask})
+        encoding.to(torch_device)
+
+        with torch.no_grad():
+            outputs = model(**encoding, interpolate_pos_encoding=True)
+
+        expected_shape = torch.Size((1, 2))
+        assert outputs.logits.shape == expected_shape
```
[the remark this answers is not in the document]
Comment (bhuvanmdev) [reply tests/models/tvp/test_modeling_tvp.py:301]: just a clarification, the model does not return the last hidden state by default.
Should I..
1.  Add `last_hidden_state` parameter in `TvpVideoGroundingOutput` in order to obtain the last_hidden_state.
 OR
2. Use `output_hidden_states=True` and do `assert outputs.hidden_states[-1].shape == expected_shape`

File: src/transformers/models/tvp/modeling_tvp.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/models/tvp/modeling_tvp.py:232]: another option - replace with `(batch_size, height, width, hidden_dim)`

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this! 

Please make sure to either comment on suggestions why you're not implementing them or implement the suggestion before marking as resolved. Comments making suggestions to make e.g. docstrings in-line with the library standard will have to be implemented

**Comment by bhuvanmdev** [comment]
@amyeroberts Done.

File: src/transformers/models/tvp/modeling_tvp.py [hunk 5]
```diff
@@ -576,6 +626,9 @@ def _init_weights(self, module):
 
         return_dict (`bool`, *optional*):
             Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+
+        interpolate_pos_encoding (`bool`, *optional*, defaults to `False`):
+            Whether to interpolate the pre-trained image pad prompter encodings.
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:631]: This should refer to the positional and pad encoding here. 

Normally, I'd say we should have the input argument reflect the behaviour better e.g. `interpolate_encoding`. However, it's probably better to be consistent across the models' forward passes for the main class.

File: src/transformers/models/tvp/modeling_tvp.py [hunk 6]
```diff
@@ -660,20 +712,53 @@ def __init__(self, config):
             )
         )
 
-    def forward(self, pixel_values):
+    def interpolate_pad_encoding(self, prompt: torch.Tensor, height: int, width: int) -> torch.Tensor:
+        """
+        This method allows to interpolate the pre-trained pad weights , to be able to use the model on collection of high
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:717]:
```suggestion
        This method allows to interpolate the pre-trained pad weights, to be able to use the model on collection of high
```

File: src/transformers/models/tvp/modeling_tvp.py [hunk 7]
```diff
@@ -660,20 +712,53 @@ def __init__(self, config):
...
+            prompt,
+            scale_factor=(h0, w0),
+            mode="bicubic",
+            align_corners=False,
+        )
+        # reversing back to (batch,frames,channels,height,width), where height and width is the new interpolated height and width
+        prompt = prompt.reshape(batch, num_frames, channels, height, width)
+        return prompt
+
+    def forward(self, pixel_values, interpolate_pos_encoding: bool = False):
```
Comment (amyeroberts) [inline src/transformers/models/tvp/modeling_tvp.py:739]: For the pad class, the argument should be `interpolate_pad_encoding` everywhere

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this! 

Just a few final things to address before merge

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this!

**Pull request merged**